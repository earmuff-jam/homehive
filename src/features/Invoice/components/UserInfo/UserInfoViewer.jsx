import React from "react";

import { Container, Stack, Typography } from "@mui/material";
import AButton from "common/AButton";
import TextFieldWithLabel from "common/TextFieldWithLabel";

export default function UserInfoViewer({
  title,
  caption,
  register,
  errors,
  isDisabled,
  onSubmit,
  loading,
}) {
  return (
    <Container
      maxWidth="sm"
      sx={{
        backgroundColor: "background.paper",
        borderRadius: 2,
        padding: 3,
        boxShadow: "0 4px 12px rgba(0, 0, 0, 0.1)",
      }}
    >
      <Stack spacing={2}>
        <Stack>
          <Typography variant="h5" fontWeight="bold">
            {title}
          </Typography>
          <Typography variant="subtitle2">{caption}</Typography>
        </Stack>

        {/* First and Last Name */}
        <Stack direction="row" spacing={2}>
          <TextFieldWithLabel
            label="First Name *"
            id="first_name"
            placeholder="First Name"
            errorMsg={errors.first_name?.message}
            inputProps={{
              ...register("first_name", {
                required: "First Name is required",
                minLength: {
                  value: 3,
                  message: "First Name must be at least three characters",
                },
                maxLength: {
                  value: 150,
                  message: "First Name must be less than 150 characters",
                },
              }),
            }}
          />
          <TextFieldWithLabel
            label="Last Name *"
            id="last_name"
            placeholder="Last Name"
            errorMsg={errors.last_name?.message}
            inputProps={{
              ...register("last_name", {
                required: "Last Name is required",
                minLength: {
                  value: 3,
                  message: "Last Name must be at least three characters",
                },
                maxLength: {
                  value: 150,
                  message: "Last Name must be less than 150 characters",
                },
              }),
            }}
          />
        </Stack>

        {/* Email and Phone Number */}
        <Stack direction="row" spacing={2}>
          <TextFieldWithLabel
            label="Email address *"
            id="email_address"
            placeholder="Email Address"
            errorMsg={errors.email_address?.message}
            inputProps={{
              ...register("email_address", {
                required: "Email Address is required",
                pattern: {
                  value: /^[^\s@]+@[^\s@]+\.[^\s@]+$/,
                  message: "Enter a valid email address",
                },
              }),
            }}
          />
          <TextFieldWithLabel
            label="Phone Number *"
            id="phone_number"
            placeholder="Phone Number"
            errorMsg={errors.phone_number?.message}
            inputProps={{
              ...register("phone_number", {
                required: "Phone number is required",
                pattern: {
                  value: /^\+?[\d\s\-().]{10,15}$/,
                  message: "Enter a valid phone number",
                },
              }),
            }}
          />
        </Stack>

        {/* Street Address */}
        <TextFieldWithLabel
          label="Street Address *"
          id="street_address"
          placeholder="Street Address"
          errorMsg={errors.street_address?.message}
          inputProps={{
            ...register("street_address", {
              required: "Street address is required",
            }),
          }}
        />

        {/* City, State, Zip Code */}
        <Stack direction="row" spacing={2}>
          <TextFieldWithLabel
            label="City *"
            id="city"
            placeholder="City"
            errorMsg={errors.city?.message}
            inputProps={{
              ...register("city", {
                required: "City is required",
              }),
            }}
          />

          <TextFieldWithLabel
            label="State *"
            id="state"
            placeholder="State"
            errorMsg={errors.state?.message}
            inputProps={{
              ...register("state", {
                required: "State is required",
                pattern: {
                  value: /^[A-Za-z]{2}$/,
                  message: "Enter a valid 2-letter state code",
                },
              }),
            }}
          />

          <TextFieldWithLabel
            label="Zip Code *"
            id="zipcode"
            placeholder="Zip Code"
            errorMsg={errors.zipcode?.message}
            inputProps={{
              ...register("zipcode", {
                required: "Zipcode is required",
                pattern: {
                  value: /^\d{5}(-\d{4})?$/,
                  message: "Enter a valid zipcode",
                },
              }),
            }}
          />
        </Stack>
        <AButton
          label="Save"
          variant="contained"
          onClick={onSubmit}
          disabled={isDisabled}
          loading={loading}
        />
      </Stack>
    </Container>
  );
}
