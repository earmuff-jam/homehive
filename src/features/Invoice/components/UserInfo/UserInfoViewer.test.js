import React from "react";
import { render, screen } from "@testing-library/react";
import UserInfoViewer from "./UserInfoViewer";

// Mock the common components used inside
jest.mock("common/AButton", () => ({
  __esModule: true,
  default: ({ label, onClick, disabled }) => (
    <button onClick={onClick} disabled={disabled}>
      {label}
    </button>
  ),
}));

jest.mock("common/TextFieldWithLabel", () => ({
  __esModule: true,
  default: ({ label, id, placeholder, errorMsg }) => (
    <div data-testid={id}>
      <label>{label}</label>
      <input placeholder={placeholder} />
      {errorMsg && <span>{errorMsg}</span>}
    </div>
  ),
}));

describe("UserInfoViewer component", () => {
  const mockRegister = jest.fn(() => ({}));
  const mockErrors = {};
  const mockSubmit = jest.fn();

  it("renders correctly and matches snapshot", () => {
    const { asFragment } = render(
      <UserInfoViewer
        title="Sender Information"
        caption="Add details about the sender"
        register={mockRegister}
        errors={mockErrors}
        isDisabled={false}
        onSubmit={mockSubmit}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders title and caption text", () => {
    render(
      <UserInfoViewer
        title="Sender Information"
        caption="Add details about the sender"
        register={mockRegister}
        errors={mockErrors}
        isDisabled={false}
        onSubmit={mockSubmit}
      />
    );

    expect(screen.getByText("Sender Information")).toBeInTheDocument();
    expect(screen.getByText("Add details about the sender")).toBeInTheDocument();
    expect(screen.getByText("Save")).toBeInTheDocument();
  });
});
