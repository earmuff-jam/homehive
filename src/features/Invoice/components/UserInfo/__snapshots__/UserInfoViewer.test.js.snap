// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserInfoViewer component renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthSm css-8gc8gz-MuiContainer-root"
  >
    <div
      class="MuiStack-root css-1sazv7p-MuiStack-root"
    >
      <div
        class="MuiStack-root css-nen11g-MuiStack-root"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 css-stnzmu-MuiTypography-root"
        >
          Sender Information
        </h5>
        <h6
          class="MuiTypography-root MuiTypography-subtitle2 css-1ypmsg6-MuiTypography-root"
        >
          Add details about the sender
        </h6>
      </div>
      <div
        class="MuiStack-root css-pl8gqh-MuiStack-root"
      >
        <div
          data-testid="first_name"
        >
          <label>
            First Name *
          </label>
          <input
            placeholder="First Name"
          />
        </div>
        <div
          data-testid="last_name"
        >
          <label>
            Last Name *
          </label>
          <input
            placeholder="Last Name"
          />
        </div>
      </div>
      <div
        class="MuiStack-root css-pl8gqh-MuiStack-root"
      >
        <div
          data-testid="email_address"
        >
          <label>
            Email address *
          </label>
          <input
            placeholder="Email Address"
          />
        </div>
        <div
          data-testid="phone_number"
        >
          <label>
            Phone Number *
          </label>
          <input
            placeholder="Phone Number"
          />
        </div>
      </div>
      <div
        data-testid="street_address"
      >
        <label>
          Street Address *
        </label>
        <input
          placeholder="Street Address"
        />
      </div>
      <div
        class="MuiStack-root css-pl8gqh-MuiStack-root"
      >
        <div
          data-testid="city"
        >
          <label>
            City *
          </label>
          <input
            placeholder="City"
          />
        </div>
        <div
          data-testid="state"
        >
          <label>
            State *
          </label>
          <input
            placeholder="State"
          />
        </div>
        <div
          data-testid="zipcode"
        >
          <label>
            Zip Code *
          </label>
          <input
            placeholder="Zip Code"
          />
        </div>
      </div>
      <button>
        Save
      </button>
    </div>
  </div>
</DocumentFragment>
`;
